src/php/lib/Grpc/BaseStub.php

```diff
@@ -203,10 +213,122 @@ private function _validate_and_normalize_metadata($metadata)
+                $options);
+            $jwt_aud_uri = $this->_get_jwt_aud_uri($method);
+            if (is_callable($this->update_metadata)) {
+                $metadata = call_user_func($this->update_metadata,
+                $metadata,
+                $jwt_aud_uri);
+            }
+            $metadata = $this->_validate_and_normalize_metadata(
+                $metadata);
+            $call->start($argument, $metadata, $options);
+            return $call;
+        };
+    }
 
-    /* This class is intended to be subclassed by generated code, so
-     * all functions begin with "_" to avoid name collisions. */
+    private function GrpcStreamUnary($channel) {
+        $this->current_channel = $channel;
+        return function ($method, $deserialize, array $metadata = [], array $options = []) {
+            $call = new ClientStreamingCall($this->current_channel,
+                $method,
+                $deserialize,
+                $options);
+            $jwt_aud_uri = $this->_get_jwt_aud_uri($method);
+            if (is_callable($this->update_metadata)) {
+                $metadata = call_user_func($this->update_metadata,
+                $metadata,
+                $jwt_aud_uri);
+            }
+            $metadata = $this->_validate_and_normalize_metadata(
+                $metadata);
+            $call->start($metadata);
+            return $call;
+        };
+    }
+
+  private function GrpcUnaryStream($channel) {
+      $this->current_channel = $channel;
+      return function ($method, $argument, $deserialize, array $metadata = [], array $options = []){
+          $call = new ServerStreamingCall($this->current_channel,
+              $method,
+              $deserialize,
+              $options);
+          $jwt_aud_uri = $this->_get_jwt_aud_uri($method);
+          if (is_callable($this->update_metadata)) {
+              $metadata = call_user_func($this->update_metadata,
+              $metadata,
+              $jwt_aud_uri);
+          }
+          $metadata = $this->_validate_and_normalize_metadata(
+              $metadata);
+          $call->start($argument, $metadata, $options);
+          return $call;
+      };
+  }
+
+  private function GrpcStreamStream($channel) {
+      $this->current_channel = $channel;
+      return function ($method, $deserialize, array $metadata = [], array $options = []){
+          $call = new BidiStreamingCall($this->current_channel,
+              $method,
+              $deserialize,
+              $options);
+          $jwt_aud_uri = $this->_get_jwt_aud_uri($method);
+          if (is_callable($this->update_metadata)) {
+              $metadata = call_user_func($this->update_metadata,
+              $metadata,
+              $jwt_aud_uri);
+          }
+            $metadata = $this->_validate_and_normalize_metadata(
+              $metadata);
+            $call->start($metadata);
+
+          return $call;
+      };
+  }
 
+  private function createCallFactory($handler, $channel){
```

Comment:
Wouldn't it be cleaner if this was split into four different functions and the argument passing/switch case was eliminated?

Reply:
Done. Yeah split can also save a lot of operation for switch during tons of unary call.